recipes/dev/early_exit_finetune_distributed.py

```diff
@@ -0,0 +1,1068 @@
+            return optimizer
+
+    def _setup_data(
+        self,
+        cfg_dataset: DictConfig,
+        shuffle: bool,
+        batch_size: int,
+        collate_fn: str,
+    ) -> Tuple[DistributedSampler, DataLoader]:
+        """
+        All data related setup happens here. Currently this recipe only supports the
+        DistributedSamplers with Map-style Datasets which fit into memory. Other samplers,
+        iterable datasets and streaming datasets are not supported.
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        if isinstance(cfg_dataset, ListConfig):
+            datasets = [
+                config.instantiate(single_cfg_dataset, self._tokenizer)
+                for single_cfg_dataset in cfg_dataset
+            ]
+            ds = ConcatDataset(datasets=datasets)
+            packed = False
+        else:
+            ds = config.instantiate(cfg_dataset, self._tokenizer)
+            packed = cfg_dataset.get("packed", False)
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
+        collate_fn = _get_component_from_path(collate_fn)
+
+        sampler = DistributedSampler(
+            ds, num_replicas=world_size, rank=rank, shuffle=shuffle, seed=0
+        )
+        dataloader = DataLoader(
+            dataset=ds,
+            batch_size=batch_size,
+            sampler=sampler,
+            # dropping last avoids shape issues with compile + flex attention
+            drop_last=True,
+            collate_fn=(
+                partial(
+                    collate_fn,
+                    padding_idx=self._tokenizer.pad_id,
+                    ignore_idx=self._loss_fn.ignore_index,
+                )
+                if not packed
+                else padded_collate_packed
+            ),
+        )
+
+        if self._is_rank_zero:
+            log.info("Dataset and Sampler are initialized.")
+
+        return sampler, dataloader
+
+    def _setup_early_exit_loss(
+        self,
+        cfg_early_exit_loss: DictConfig,
+    ) -> Tuple[List[bool], EarlyExitCurriculum]:
+        """
+        All early exit loss related setup happens here.
+        """
+        do_output_hidden_states = None
+        early_exit_loss_curriculum = None
+
+        if cfg_early_exit_loss:
+            assert (
+                hasattr(self._loss_fn, "reduction")
+                and self._loss_fn.reduction == "mean"
+            ), "Currently early exit loss is only implemented for loss functions that apply a mean reduction."
+
+            do_output_hidden_states = slice_str_to_array(
+                cfg_early_exit_loss.get("layers", ":"), len(self._model.layers)
+            )
+            train_last_layer = cfg_early_exit_loss.get("include_last_layer", True)
+            verbose = cfg_early_exit_loss.get("verbose", False)
+
+            if cfg_early_exit_loss.curriculum:
```

Comment:
I think it's possible that `cfg_early_exit_loss` may not have the `curriculum` attribute

Reply:
Oh the usage should be with the function/class you want to instantiate as part of the config (passed via the `_component_` field). So not too dissimilar to Hydra instantiate but `_target_` -> `_component_` (and we don't support nested instantiation or other advanced Hydra features).

So you would need to add the following to your config:

```
early_exit_loss:
  _component_: torchtune.modules.early_exit_loss.setup_early_exit_loss_curriculum
  ... # other existing fields
```

Then call as 

```
early_exit_loss_curriculum = config.instantiate(cfg_early_exit_loss, ...)
```

But this works best if the fields in your config are simply a passthrough to the function you're instantiating. Since you're modifying things in the recipe (e.g. renaming `curriculum` -> `early_exit_curriculum`, using a wrapper to cast the `layers` field into a list) you would need to change these behaviors. (Note that passing additional derived args like `max_steps` is fine as `instantiate` supports that. The bigger concern is that you do not have erroneous config keys floating around)

Reply:
OK. I see, since we do some processing/casting rather than a simple passthrough then this might require some refactoring of the existing implementation. So you think we can have it in a future PR?